Add users

lindavin opened the pull request:
# Description

This PR adds user roles into OED. It adds the following user roles: Admin, Obvius, Csv, and Export to regulate usage within OED. For user role requirements, we add the:
- Admin role requirement where the user wants to POST via the groups, maps, meters, preferences, users API.
- Obvius role requirement on the obvius API.
- Csv role requirement on the fileProcessing API.
- Export role requirement on the export API.
 
We also added an interface for Admins to create new users, delete users, and modify user roles.

Fixes #544

## Type of change

- [ ] Note merging this changes the node modules
- [x] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request

## Limitations

- This PR does not provide a direct way for Admins to reset user passwords, which corresponds to issue #265. However, it can be indirectly achieved by creating new users and deleting old ones. For example, to change the test user's password, an Admin would login as the test user and create a new Admin user. Then they would log in with the new Admin account and delete the old test account.
- Future test code may want a set of tests that ensure that other user roles can perform a specific action. For instance, to ensure that only Admins can delete a user we add mocha tests that loop over the other user roles to perform the request and observe a failure (status code 403). It would be better if this could be generalized into one reusable function.
- We hard code the 403 HTTP Status code to indicate an insufficient role. It would be better if these status codes were standardized somewhere in OED.

lindavin committed:
Add role to Users table

lindavin committed:
Change roles column in Users table and include it in the test user

lindavin committed:
Update retrieval to include role

lindavin committed:
Add role to user modification

lindavin committed:
Include row when retrieving all users

lindavin committed:
Include role in the login process

lindavin committed:
Fix mistake in param

We applied the role param to the password_hash param by mistake.

lindavin committed:
Add default csv and obvius users on install

lindavin committed:
Move role into jwt payload

lindavin committed:
Adds middleware that valids admin status

lindavin committed:
Move middleware for admin verification

lindavin committed:
Fix typo

lindavin committed:
Add validation to obvius route

lindavin committed:
Add function that checks if a token is authorized for a role

lindavin committed:
Add function that checks if credentials are authorized

lindavin committed:
The isTokenAuthorized should only be used on a token that has already been verified

It may have to change to use the asynchrounous jsonwebtoken#verify method. We will try to promisify that method so it is easier to use.

lindavin committed:
Rewrite validation functions

lindavin committed:
Add User type for React

lindavin committed:
Add users view and route

lindavin committed:
Add stub component for creating users

lindavin committed:
Add buttons to access different user management actions

lindavin committed:
Add route to creating a user

lindavin committed:
Convert formatting and remove unused component

lindavin committed:
Add api to retrieve users

lindavin committed:
Connect component to api

lindavin committed:
Add component that allows admin to create users

lindavin committed:
Connect component to the server so admins can create users

lindavin committed:
Update comments

lindavin committed:
Make user management page consistent with others

The user management interface is now consistent with the pages. The link to it
will only be found in the adming page (user page) and the link to creating a
new user will be located in the details/edit page.

lindavin committed:
Add role enum in database

lindavin committed:
Style create user form

lindavin committed:
Add ADMIN role to test user

lindavin committed:
Fix bugs with creating users

Return response to client and add salt rounds to bcrypt.

lindavin committed:
Redirect to users view

lindavin committed:
Format code

lindavin committed:
Add stub button for user edits

lindavin committed:
Fix validation errors and add tests

lindavin committed:
Add test for creating users

lindavin committed:
Add route and tests for updating user role

lindavin committed:
Add route and tests for deleting a user

lindavin committed:
Remove unnecessary route

lindavin committed:
Change improper index

lindavin committed:
Add frontend capability to edit and delete users

lindavin committed:
Add admin level authentication and tests to user management

lindavin committed:
Add notifications to user creation

lindavin committed:
Small fix to tests

We forget to include the token when sending the chai request.

lindavin committed:
Protect user edits from losing all admins

lindavin committed:
Remove role from token to prevent issues with stale tokens

lindavin committed:
Reset history so that newer user edits are detected

lindavin committed:
Prevent admins from accidentally deleting themselves

lindavin committed:
Revert "Add default csv and obvius users on install"

This reverts commit 151dd03974585effc1d3deeb5b2205279ef99ca9.

lindavin committed:
Add export role

lindavin committed:
Reorder keys for consistency

lindavin committed:
Add comments about the roles enum

lindavin committed:
Fix bugs with obvius authentication

lindavin committed:
Add role tests for obvius route

lindavin committed:
Loop over non admin roles

lindavin committed:
Add tests for admin role

lindavin committed:
Add admin role authentication to preferences

lindavin committed:
Add tests for updating preferences according to role

lindavin committed:
Minor stlying changes to create user form

lindavin committed:
Merge remote-tracking branch 'upstream/development' into addUsers

lindavin committed:
Use clearer language

lindavin committed:
Fix npm run check errors

lindavin committed:
Add color to buttons

lindavin committed:
Fix meters role validation

lindavin committed:
Add extra check in case user was deleted

lindavin committed:
Fix meters check

lindavin committed:
Add role check to maps route

lindavin committed:
Fixing missing conn

lindavin committed:
Merge branch 'development' into addUsers

lindavin committed:
Add role check when exporting raw readings

lindavin committed:
Add general role test middleware

lindavin committed:
Clarify action

lindavin committed:
Add authentication to maps

lindavin committed:
Add authentication to groups

lindavin committed:
Add authentication to logs api

lindavin committed:
Change error code to 403

lindavin committed:
Add checks for unauthorized users

lindavin committed:
Add csv authentication

lindavin committed:
Add authentication to see users and better format user api tests

lindavin committed:
Add tests to reject unauthorized roles

lindavin committed:
Add checks for unauthorized roles

lindavin committed:
Add better structure to obvius test

lindavin committed:
Change test hierarchy to API, modification, endpoint

lindavin committed:
Fix linter errors

lindavin committed:
Merge branch 'development' into addUsers

lindavin committed:
Change footer component to container

lindavin committed:
Add proper check for export role

lindavin committed:
Only modify req when necessary

lindavin committed:
Make naming consistent

lindavin committed:
Update README with username

lindavin committed:
Move userRoles to util

lindavin committed:
Add create endpoint

lindavin committed:
Include pass_hash when retrieving by id

lindavin committed:
Add comments and clarify code

lindavin committed:
Promisify jwt#verify

lindavin committed:
Add try/catch

lindavin committed:
Merge branch 'development' into addUsers

lindavin committed:
Sanitize email

huss commented on src/client/app/utils/api/UsersApi.ts:
```diff
@@ -0,0 +1,48 @@
+interface NewUser extends User {
+	password: string;
+}
+export default class UsersApi {
```
Should have blank line before class.

lindavin replied:
Thanks!

huss commented on src/server/models/User.js:
```diff
@@ -83,6 +107,28 @@ class User {
+}
+
+/**
+ * If enum needs to be kept in sync with the src/server/sql/create_user_types_enum.sql and the UserRoles enum in src/client/types/items.ts 
```
Is this "If enum" or "enum" since I think this should be true.

lindavin replied:
Thanks for the catch.

huss commented on src/server/routes/authenticator.js:
```diff
@@ -15,23 +21,150 @@ authMiddleware = (req, res, next) => {
+		properties: {
+			email: {
+				type: 'string',
+				minLength: 3,
```
Just being picky but wouldn't a@b.c be the min or 5 characters?

lindavin replied:
Thanks done.

huss commented on src/server/routes/authenticator.js:
```diff
@@ -15,23 +21,150 @@ authMiddleware = (req, res, next) => {
+			},
+			password: {
+				type: 'string',
+				minLength: 3
```
I think OED requires at least 8 character passwords so maybe that should be the min? I would fail later so not essential.

lindavin replied:
Thanks done.

huss commented on src/server/services/user/editUser.js:
```diff
@@ -22,6 +23,12 @@ const { getConnection, dropConnection } = require('../../db');
 		terminateReadline('Password must be at least eight characters, user\'s password not modified');
 	}
 
+	role = await ask('Role: '); // untested
```
Why does it say "untested"? Does this need to be done via a script (sorry, did not check)?

lindavin replied:
It's untested because I didn't formally check that this works (by running it or writing a mocha test). I am fairly confident that it should work as intended though. Do you think I should add a mocha test?

huss replied:
I think this is fine but the comment confused me.

huss commented on src/server/services/user/editUser.js:
```diff
@@ -11,6 +11,7 @@ const { getConnection, dropConnection } = require('../../db');
 (async () => {
```
extra blank line above (not @lindavin edit but nice to fix up).

huss replied:
fixed.

huss commented on src/server/services/user/editUser.js:
```diff
@@ -22,6 +23,12 @@ const { getConnection, dropConnection } = require('../../db');
 
+	role = await ask('Role: '); // untested
+	role = role.toUpperCase();
+	if(User.role[role] === undefined){
```
Spacing issues.

huss replied:
fixed

huss commented on src/server/test/web/groups.js:
```diff
@@ -5,9 +5,11 @@
 /* This file tests the groups API. */
 
 const { chai, mocha, expect, app, testUser, testDB } = require('../common');
+const bcrypt = require('bcryptjs');
```
I like the new tests. However, here and in other files, they tend to not check a valid request from an admin. I thought one way to do it would be to have an else on the check for the admin role in the loops. Another would be to put that check inside each check so less code needs to be repeated. Just and idea of how to code and don't really care how it happens.

lindavin replied:
The testUser imported on line 7 is an admin by default so the other checks are an implicit check on the admin role. However, I can make an explicit check for the admin role. I like the first suggestion of the else check. I don't see how putting that check inside each check would mean that less code is repeated. Wouldn't I be adding a line to each code so it's more repetition? I'll think about it some more, but will probably go with the first suggestion.

lindavin replied:
I have taken the first suggestion for groups.js. For the other tests, I rewrote 'authorized role' -> 'admin role' to be more explicit. Let me know if this is okay; I put this work in the commit 'Write explicit tests for Admin role'.

huss commented on src/server/test/web/usersTest.js:
```diff
@@ -0,0 +1,121 @@
+const { chai, mocha, expect, app, testDB, testUser } = require('../common');
+const User = require('../../models/User');
+const bcrypt = require('bcryptjs');
+mocha.describe('Users API', () => {
```
Blank line between code and import.

lindavin replied:
Thanks!

huss commented on src/server/util/userRoles.js:
```diff
@@ -0,0 +1,47 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const jwt = require('jsonwebtoken');
```
Is the jwt import needed?

lindavin replied:
Yes. We use the promisify package to turn jwt#verify into a promise.

huss replied:
Sorry, my mistake. I missed the other reference.

huss commented on src/server/util/userRoles.js:
```diff
@@ -0,0 +1,47 @@
+const { getConnection } = require('../db');
+const secretToken = require('../config').secretToken;
+
+/** Checks if a role has the authorization capabilities as the requested role. */
```
All the functions should have JSDoc with @param and @return. Esp. important since others are likely to use these.

lindavin replied:
Thanks done.

huss commented on src/server/util/userRoles.js:
```diff
@@ -0,0 +1,47 @@
+ * 
+ * @param {User} user 
+ * @param requestedRole 
+ */
```
Missing return and the comment "description" should be within JSDoc as a single comment.

lindavin replied:
Thanks!

huss reviewed (changes requested):
This is good work and implements what is needed. A few general comments:

- When I log in in a role other than admin, I can still get to the admin, user, meter, group & map page. They show all the same items/table headers as when in the admin role but it fails when you try to do a save of changes. Do I have an issue with my install or is this a more general issue? Note if not logged in then the pages appear as expected.
- There are 5 CodeQL tests that fail. These are due to not rate limiting. This is a larger issue with OED that needs to be addressed outside this PR. Hopefully we can easily disable these until the needed changes are made so PRs will stay okay for the checks.
- I still need to check that Obvius, csv & export does what is expected in those roles but am waiting on that.

lindavin committed:
Add missing blank lines

lindavin committed:
Minor changes to comments

lindavin committed:
Fix min length issues

lindavin committed:
Add params and return to functions

lindavin committed:
Add references to translations

lindavin committed:
Write explicit tests for Admin role

lindavin commented:
I have added resolved most of the comments from the review by @huss. I setup internationalization of the text, but we will need to have actual translations. I marked the text that need to be translated with a "(Needs <language>)" in the data.json. As for testing checking that a request from the admin is valid, I have now added a test for than in groups.js.  Additionally, since the testUser is an admin user, the existing tests should now fill the role of checking a valid request from an Admin.

huss committed:
fix spacing issues

lindavin commented:
Regarding the issue of failing when you try to do a save of changes when not an admin, this is working as expected because only admins should. Should the interface change so that you cannot get certain pages if you are not an admin? I allowed access to those pages because ultimately the user cannot save any changes unless they are an admin.

huss commented:
> Regarding the issue of failing when you try to do a save of changes when not an admin, this is working as expected because only admins should. Should the interface change so that you cannot get certain pages if you are not an admin? I allowed access to those pages because ultimately the user cannot save any changes unless they are an admin.

I was looking at this right now. src/client/app/utils/token.ts has hasToken() and that is used in many places to decide if items should be displayed. I think in all these places the idea was that only an admin could see that info. I think we need to change them to use your new admin authentication so it does the same thing. Otherwise, anyone logged in with any role can see the info.

I understand that your thinking was anyone could get to the page but only admins could edit. However, I have two concerns with this. First, it exposes information to non-admins that they really should not see. Second, I think it can be confusing to the user. Yes, they would need to be logged in as some user but I don't think non-admins should see this info. This is what the help pages now show.

Is this something you want to do or should someone else take this on? I think it needs to be an update to this PR either way.

huss commented:
One more item. The lint checks are failing from the edit to src/server/test/web/groups.js.

lindavin committed:
Fix lint errors

lindavin committed:
Merge branch 'addUsers' of github.com:lindavin/OED into addUsers

lindavin commented:
I can take on the task of having the UI change depending on the role. Plus I want to learn a bit of redux as I don't know it. This is what I'm thinking for the implementation, I'm open to suggestions:
1) On first load (i.e. before the app mounts), check whether or not there is a valid token in `localStorage` via `checkToken`. If there is get the role via the `/users/token` route and put it into the store. Otherwise, we wait until the user has logged in to store the role into the redux store.
2) Components will check the store to see if the user is an admin and change accordingly.
3) If the roles has changed while the user is logged in, they would reload their page or log out then log in.

lindavin commented:
I've gone ahead with this strategy with some success.

> I can take on the task of having the UI change depending on the role. Plus I want to learn a bit of redux as I don't know it. This is what I'm thinking for the implementation, I'm open to suggestions:
> 
>     1. On first load (i.e. before the app mounts), check whether or not there is a valid token in `localStorage` via `checkToken`. If there is get the role via the `/users/token` route and put it into the store. Otherwise, we wait until the user has logged in to store the role into the redux store.
> 
>     2. Components will check the store to see if the user is an admin and change accordingly.
> 
>     3. If the roles has changed while the user is logged in, they would reload their page or log out then log in.

lindavin committed:
Fix comment

lindavin committed:
Fix security issue, do not return password hash

lindavin committed:
Fix formatting to consider admin role

lindavin committed:
Refactor permission checks into separate functions

lindavin committed:
Add role into redux store

We did this by copying and modifying the way that the version is stored
in the redux store. We include both role and email as the 'profile' as
it seemed natural to include more than just the role in the store. This
is also just in case need the email in the future, perhaps to send
emails to the logged in user.

lindavin committed:
Save profile in redux store on app load and login

lindavin committed:
Add admin check when displaying header and meters

lindavin committed:
Rename profile to current user where possible for clarity

lindavin committed:
Add admin check to groups page

lindavin committed:
Clear profile on logout

lindavin committed:
Delete and clear stale tokens and profiles

lindavin committed:
Remove trailing whitespace

lindavin committed:
Remove instances of hasToken

We remove instances of hasToken to remove the import and because the
its use in those areas is to only allow authenticated admins to access
certain information. The check that we had added serves that
purpose since with user roles, simply checking for a token is not enough
to ensure that the user is an admin. Since we had forced a direct
relationship between having a token and storing the user's profile into
the redux store, checking that we have a token is redudant.

lindavin committed:
Fix bug, logout on homepage does not change header buttons

This bug was introduced within this PR when we removed forceUpdate.

lindavin committed:
Protect client side routes using roles

lindavin committed:
Merge branch 'development' of git://github.com/OpenEnergyDashboard/OED into addUsers

lindavin committed:
Fix failing meter test

lindavin commented:
It took more work than anticipated, but now the UI appears the same to both authenticated non-Admin users and unauthenticated users. Only admins should have access to the admin, user, meter, group & map page with full information. The links to these pages are hidden when not an admin and attempts to access these pages via the url will redirect the user to the home page. When I attempted to access pages restricted to just admins as an authenticated non-Admin, I could still see the page for a split second. I do not know how to fix this. Open to suggestions, unless we are okay with that for now.

I did not replace all instances of hasToken because it is still a useful method to check whether that the user is not authenticated.

huss commented on src/client/app/utils/hasPermissions.ts:
```diff
@@ -0,0 +1,19 @@
+ * Checks if the role A has the permission of the role B.
+ */
+export function hasPermissions(A: UserRole, B: UserRole): boolean {
+	return A === UserRole.ADMIN || A === B;
```
I might add this comment:
// Admins always have any other role.

lindavin replied:
Thanks! Done.

huss commented on src/client/app/utils/hasPermissions.ts:
```diff
@@ -0,0 +1,19 @@
+/**
+ * Checks if the role A has the permission of the role B.
+ */
+export function hasPermissions(A: UserRole, B: UserRole): boolean {
```
I can see why you named these A & B but normally only classes, etc. are capitalized. Could it be either "a" or "userA" (I prefer the second one).

lindavin replied:
Thanks! Done.

huss commented on src/client/app/reducers/currentUser.ts:
```diff
@@ -0,0 +1,37 @@
+import { ActionType } from '../types/redux/actions';
+
+/*
+* Defines store interactions when version related actions are dispatched to the store.
```
Should this be profile related changes?

lindavin replied:
Thanks! Done.

huss reviewed (changes requested):
There are a couple of small comments to clear up.

As for seeing the page briefly when not an admin, I needed to manually access that page by entering the URL. This seems unlikely and no security information is displayed. I think it is okay for now.

My only other comment is that when I'm logged in as a user other than admin and go to the meters page I still see other columns of data and the column headers are off. I think this may be related to using hasToken() instead of checking for admin role in src/client/app/components/meters/MeterViewComponent.tsx. However, in src/client/app/components/meters/MetersDetailComponent.tsx it uses loggedInAsAdmin to decide what headers to put there and this seems correct.

lindavin committed:
Add and fix comments from feedback

lindavin committed:
Fix issues of view showing data for non-admin users

lindavin commented:
Thanks for the comments and the catch for the meters view! This is fixed now.

huss committed:
name column identifer not name

huss commented:
This is now ready to merge. I thank @lindavin for working on more areas than originally expected and following through to get this all done.

huss commented:
I want to leave a comment that this PR is causing CodeQL failures for rate limiting. This has been a known issue (see #266). CodeQL has suggested fixes that are not too hard so it is hoped to fix this soon. However, it will be done outside of this PR so it can move forward.
